doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
 
-These are guidelines that are generally followed:
+The first step is to `clone`_ the MNE-Python repository from your fork, and
+also connect the local copy to the ``upstream`` version of the codebase, so you
+can stay up-to-date with changes from other contributors. First, edit these two
+variables for your situation:
 
-Pull requests
-^^^^^^^^^^^^^
-* Address one issue per pull request (PR).
-* Avoid unnecessary cosmetic changes in PRs.
-* Minimize test timing while maximizing coverage. Use ``pytest --durations=20`` on modified tests.
-* Update the ``doc/whats_new.rst`` file last, just before merge to avoid merge conflicts.
+.. code-block:: console
 
-Naming
-^^^^^^
-* Classes should be named using CamelCase.
-* Functions and instances/variables should be snake_case (n_samples rather than nsamples).
-* Avoid single-character variable names.
+    $ INSTALL_LOCATION="/opt"
+    $ GITHUB_USERNAME="new_mne_contributor"
 
-Importing
-^^^^^^^^^
-* Import modules in this order:
-  1. builtin
-  2. standard scientific (``numpy as np``, ``scipy`` submodules)
-  3. others
-  4. mne imports (relative within the MNE module, absolute in the examples)
-* Imports for ``matplotlib`` and optional modules (``sklearn``, and ``pandas``, etc.) within the MNE module should be nested (i.e., within a function or method, not at the top of a file).
+Then make a local clone of your remote fork ("origin"):
 
-Vizualization
-^^^^^^^^^^^^^
-* Add public functions to the :mod:`mne.viz` package and use these in the corresponding methods.
-* All visualization functions must accept a ``show`` parameter and return a ``fig`` handle.
-* Use ``RdBu_r`` colormap for signed data with a meaningful middle (zero-point) and ``Reds`` otherwise in visualization functions and examples.
+.. code-block:: console
 
-Return types
-^^^^^^^^^^^^
-* Methods should modify inplace and return ``self``, functions should return copies (where applicable).
+    $ cd $INSTALL_LOCATION
+    $ git clone https://github.com/$GITHUB_USERNAME/mne-python.git
+
+Finally, set up a link between your local clone and the official repository
+("upstream"):
+
+.. code-block:: console
+
+    $ cd mne-python
+    $ git remote add upstream git://github.com/mne-tools/mne-python.git
+    $ git fetch --all
+
+.. note::
+
+    We use ``git://`` instead of ``https://`` in the address for the official
+    "upstream" remote repository. ``git://`` addresses are read-only, which
+    means you can *pull* the official repository into your local copy (in order
+    to stay up-to-date with changes made by other contributors) but you cannot
+    *push* anything from your computer directly into the upstream remote.
+    Instead, you must push your changes to your own remote fork first, and then
+    create a pull request from your remote into the upstream remote (and even
+    then, your changes are not automatically accepted into the upstream; the
+    changes must be *approved* by maintainers and then *merged* into upstream).
```

Comment:
If this can become collapsible, let's collapse it maybe with a title "Why https:// for one and git:// for the other?". If not, I think we can get rid of it.